### PR TITLE
asserts,interfaces/policy: add support for $SLOT()/$PLUG()/$MISSING in *-attributes constraints

### PR DESCRIPTION
**Pull request by pedronis**
This adds support for $SLOT(arg), $PLUG(arg) and $MISSING attribute constraints in plugs and slots rules in snap-declarations:

$SLOT(dotted), $PLUG(dotted) instead of constraining by a regexp, try to retrieve what is specified by dotted, a dotted chain of attribute names, from the SLOT respectively PLUG attributes, the retrieved value is what the constrained attribute must match. This allows for example to express that an attribute must be equal between slot and plug sides. If a value cannot be retrieved this constraints fail.

$MISSING constraints the attribute to be not set.

### CODE REVIEW

File: asserts/ifacedecls.go [hunk 1]
```diff
@@ -78,6 +85,12 @@ func compileAttrMatcher(cc compileContext, constraints interface{}) (attrMatcher
 		if !cc.hadMap {
 			return nil, fmt.Errorf("first level of non alternative constraints must be a set of key-value contraints")
 		}
+		if strings.HasPrefix(x, "$") {
+			if x == "$MISSING" {
+				return missingAttrMatcher{}, nil
+			}
+			return compileEvalAttrMatcher(cc, x)
+		}
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Logic here gives a pretty good feeling. Small spot clearly returning the alternatives for handling different rule styles.

File: asserts/ifacedecls.go [hunk 2]
```diff
@@ -98,42 +111,94 @@ func compileMapAttrMatcher(cc compileContext, m map[string]interface{}) (attrMat
 	return matcher, nil
 }
 
-func matchEntry(apath, k string, matcher1 attrMatcher, v interface{}) error {
+func matchEntry(apath, k string, matcher1 attrMatcher, v interface{}, ctx AttrMatchContext) error {
 	apath = chain(apath, k)
-	if v == nil {
+	if _, ok := matcher1.(missingAttrMatcher); !ok && v == nil {
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Shouldn't that be done in the other (non-missing) constraint implementation themselves?

File: asserts/ifacedecls.go [hunk 3]
```diff
@@ -98,42 +111,94 @@ func compileMapAttrMatcher(cc compileContext, m map[string]interface{}) (attrMat
...
+
+func (matcher missingAttrMatcher) match(apath string, v interface{}, ctx AttrMatchContext) error {
+	if v != nil {
+		return fmt.Errorf("attribute %q is constrained to be missing but is set", apath)
+	}
+	return nil
+}
+
+type evalAttrMatcher struct {
+	// first iteration supports just $SLOT|PLUG(argument1)
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: (SLOT|PLUG) might read a bit more clearly, but trivial nitpicking obviously.

File: asserts/ifacedecls.go [hunk 4]
```diff
@@ -98,42 +111,94 @@ func compileMapAttrMatcher(cc compileContext, m map[string]interface{}) (attrMat
...
+var (
+	validEvalAttrMatcher = regexp.MustCompile(`^\$(SLOT|PLUG)\((.+)\)$`)
+)
+
+func compileEvalAttrMatcher(cc compileContext, s string) (attrMatcher, error) {
+	ops := validEvalAttrMatcher.FindStringSubmatch(s)
+	if len(ops) == 0 {
+		return nil, fmt.Errorf("cannot compile %q constraint %q: not a valid $SLOT()/$PLUG() constraint", cc, s)
+	}
+	// XXX: prohibit $SLOT() in slot-attributes and vice versa?
```
Comment (niemeyer) [inline asserts/ifacedecls.go]: Not sure.. seems useful to cross-reference. If we can't see a good reason to leave out, might be nice.

File: asserts/ifacedecls.go (hunk 2 above)
Comment (pedronis) [inline asserts/ifacedecls.go]: we could, but given that we decided that only $MISSING should be an exception to the rule that if attribute has a constraint, attribute must be present, I thought it's better to express the exception this way (I could put a clear comment about that though), we can also revisit this once $MISSING is not the only exception anymore for that

**Comment by pedronis** [comment]
the assertion service has some (in retrospect) silly code that is preventing the tests to pass, we are trying to deploy a fix

File: asserts/ifacedecls.go [hunk 5]
```diff
@@ -98,42 +111,94 @@ func compileMapAttrMatcher(cc compileContext, m map[string]interface{}) (attrMat
...
-			if err := matchEntry(apath, k, matcher1, x[k]); err != nil {
+			if err := matchEntry(apath, k, matcher1, x[k], ctx); err != nil {
 				return err
 			}
 		}
 	case []interface{}:
-		return matchList(apath, matcher, x)
+		return matchList(apath, matcher, x, ctx)
 	default:
 		return fmt.Errorf("attribute %q must be a map", apath)
```
Comment (chipaca) [inline asserts/ifacedecls.go]: ...except when it can be a list?

**Pull request merged**